projects/packages/plugin-deactivation/src/class-deactivation-handler.php

```diff
@@ -0,0 +1,117 @@
+
+	/**
+	 * Plugin Deactivation package version
+	 *
+	 * @var string
+	 */
+	const PACKAGE_VERSION = '0.1.0-alpha';
+
+	/**
+	 * Slug of the plugin to intercept deactivation for.
+	 *
+	 * @var string
+	 */
+	private $plugin;
+
+	/**
+	 * Path to a PHP file that will be used as a template for the deactivation dialog.
+	 *
+	 * Copy ./dialog-template.php to your plugin and modify it to suit your needs.
+	 *
+	 * @var string
+	 */
+	private $dialog_view;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string $plugin       Slug of the plugin to intercept deactivation for.
+	 * @param string $dialog_view  Path to a PHP file that will be used as a template for the deactivation dialog.
+	 */
+	public function __construct( $plugin, $dialog_view ) {
+		$this->plugin      = $plugin;
+		$this->dialog_view = $dialog_view;
+	}
+
+	/**
+	 * Instantiates the deactivation handler to intercept the deactivation of plugins.
+	 *
+	 * @param string $plugin       Slug of the plugin to intercept deactivation for.
+	 * @param string $dialog_view  Path to a PHP file that will be used as a template for the deactivation dialog.
+	 */
+	public static function init( $plugin, $dialog_view ) {
+		$instance = new self( $plugin, $dialog_view );
+
+		if ( ! file_exists( $instance->dialog_view ) ) {
+			wp_die( 'The plugin deactivation dialog view file does not exist.' );
+		}
+
+		add_action( 'load-plugins.php', array( $instance, 'enqueue_script' ) );
+		add_action( 'admin_footer-plugins.php', array( $instance, 'embed_dialog' ), 9 );
+		add_action( 'admin_footer-plugins.php', array( $instance, 'init_dialog' ), 10 );
+
+		return $instance;
+	}
+
+	/**
+	 * Enqueues the deactivation handler script and styles.
+	 */
+	public function enqueue_script() {
+		wp_enqueue_script(
+			'jp-plugin-deactivation',
+			plugins_url( '../dist/deactivation.js', __FILE__ ),
+			array(),
+			self::PACKAGE_VERSION,
+			true
+		);
+
+		wp_enqueue_style(
+			'jp-plugin-deactivation',
+			plugins_url( '../dist/deactivation.css', __FILE__ ),
+			array(),
+			self::PACKAGE_VERSION
+		);
+	}
+
+	/**
+	 * Add required html to the plugin page footer.
+	 */
+	public function embed_dialog() {
+		?>
+		<div id="jp-plugin-deactivation-<?php echo esc_attr( $this->plugin ); ?>" class="jp-plugin-deactivation">
+			<div class="jp-plugin-deactivation__dialog">
+				<?php include $this->dialog_view; ?>
+			</div>
+			<div 
+				data-jp-plugin-deactivation-action="close"
+				class="jp-plugin-deactivation__overlay"
+			></div>
+		</div>
+		<?php
+	}
+
+	/**
+	 * Instantiate the dialog handler class.
+	 */
+	public function init_dialog() {
+		// Name of the javascript variable that will hold the dialog handler instance for a given plugin.
+		$variable_name = str_replace( '-', '', ucwords( sanitize_key( $this->plugin ), '-' ) ) . 'Deactivation';
+
+		echo '<script>var ' . esc_js( $variable_name ) . ' = new JetpackPluginDeactivation( decodeURIComponent( "' . rawurlencode( $this->plugin ) . '" ) );</script>';
```

Comment:
For this here, I think we need to leverage `wp_localize_script`. I know it's counter intuitive, but `esc_js` doesn't actually escape for script tags, it escapes content for inline javascript attributes: https://developer.wordpress.org/reference/functions/esc_js/

A better solution here would look something like this:

Gather all the plugins that have been registered with `Deactivation_Handler` class in an array, something like:
```php
$plugins = [ 'jetpack', 'jetpack-boost' ];
```

And then pass those to `jp-plugin-deactivation` using `wp_localize_script` and let WordPress worry about data sanitization.

Then, in `init_dialog`, read those values from where `wp_localize_script` stored them, something like:

```js
window.jetpack_plugin_deactivation_instances = window.jpd_plugin_slug_list.map(
  ( slug ) => new JetpackPluginDeactivation( slug )
)
```

Reply:
It uses `sanitize_key` before coming to `esc_js`. If we used the `wp_localize_script` approach, we would need to run the `init_dialog` function only once. Which has a couple of problems:
1. Consumers cannot control the instantiation individually by overriding the method, which they can do now. Keeping the instantiation separate for each plugin keeps it flexible and simple. e.g. In case we need to pass configuration parameters to customize the behavior as you suggested in one of your earlier comments, we can do that very easily when instantiation is individual.
2. Since it runs only once, the method would probably have to be static.

Reply:
This seems safe to me.